service/src/main/java/com/codeforcommunity/processor/ReservationProcessorImpl.java

```diff
@@ -0,0 +1,233 @@
+package com.codeforcommunity.processor;
+
+import static org.jooq.generated.Tables.*;
+
+import com.codeforcommunity.api.IReservationProcessor;
+import com.codeforcommunity.auth.JWTData;
+import com.codeforcommunity.dto.reservation.*;
+import com.codeforcommunity.enums.PrivilegeLevel;
+import com.codeforcommunity.enums.ReservationAction;
+import com.codeforcommunity.exceptions.*;
+import java.sql.Timestamp;
+import java.util.Optional;
+import org.jooq.DSLContext;
+import org.jooq.generated.tables.records.*;
+
+public class ReservationProcessorImpl implements IReservationProcessor {
+
+  private final DSLContext db;
+
+  public ReservationProcessorImpl(DSLContext db) {
+    this.db = db;
+  }
+
+  public Optional<ReservationsRecord> lastAction(int blockId) {
+    return Optional.ofNullable(
+        db.selectFrom(RESERVATIONS)
+            .where(RESERVATIONS.BLOCK_ID.eq(blockId))
+            .orderBy(RESERVATIONS.PERFORMED_AT.desc())
+            .limit(1)
+            .fetchOne());
+  }
+
+  private void basicChecks(int blockId, Integer userId, Integer teamId) {
+    if (!db.fetchExists(db.selectFrom(BLOCKS).where(BLOCKS.ID.eq(blockId)))) {
+      throw new ResourceDoesNotExistException(blockId, "block");
+    }
+
+    if (userId != null) {
+      if (!db.fetchExists(db.selectFrom(USERS).where(USERS.ID.eq(userId)))) {
+        throw new UserDoesNotExistException(userId);
+      }
+    }
+
+    if (teamId != null) {
+      if (!db.fetchExists(db.selectFrom(TEAMS).where(TEAMS.ID.eq(teamId)))) {
+        throw new ResourceDoesNotExistException(teamId, "team");
+      }
+    }
+
+    if (userId != null && teamId != null) {
+      if (!db.fetchExists(
+          db.selectFrom(USERS_TEAMS)
```

Comment:
Do we still have the idea of pending user_team requests? If so, we need to exclude them from this db query